src/GraphQL/Internal/Validation.hs

```diff
@@ -208,18 +192,165 @@ validateOperation (Mutation vars directives selectionSet) = do
+  unresolved <- lift (traverse validateSelection selections)
+  resolved <- traverse (resolveSelection fragments) unresolved
+  lift $ groupByResponseKey resolved
 
--- | Turn a set of arguments from the AST into a guaranteed unique set of arguments.
+-- | A selection set, almost fully validated.
 --
--- <https://facebook.github.io/graphql/#sec-Argument-Uniqueness>
-validateArguments :: [AST.Argument] -> Validation (Arguments AST.Value)
-validateArguments args = Arguments <$> mapErrors DuplicateArgument (makeMap [(name, value) | AST.Argument name value <- args])
+-- Sub-selection sets might not be validated.
+newtype SelectionSet value = SelectionSet (OrderedMap ResponseKey (ExecutionField value)) deriving (Eq, Ord, Show)
+
+newtype SelectionSetByType value
+  = SelectionSetByType (OrderedMap ResponseKey (OrderedMap (Set TypeCondition) (ExecutionField value)))
+  deriving (Eq, Ord, Show, Functor, Foldable, Traversable)
+
+-- | A 'ResponseKey' is the key under which a field appears in a response. If
+-- there's an alias, it's the alias, if not, it's the field name.
+type ResponseKey = Name
+
+-- | A field ready to be executed.
+-- XXX: Reviewer, the 'Field' name has been freed up. Should I rename this to 'Field'.
+data ExecutionField value
+  = ExecutionField
+  { name :: Name
+  , arguments :: Arguments value
+  , subSelectionSet :: Maybe (SelectionSetByType value)
+  } deriving (Eq, Ord, Show, Functor, Foldable, Traversable)
+
+instance HasName (ExecutionField value) where
+  getName = name
+
+-- | Get the value of an argument in a field.
+lookupArgument :: ExecutionField value -> Name -> Maybe value
+lookupArgument (ExecutionField _ (Arguments args) _) name = Map.lookup name args
+
+-- | Get the selection set within a field.
+getSubSelectionSet :: ExecutionField value -> Maybe (SelectionSetByType value)
+getSubSelectionSet = subSelectionSet
+
+-- | Merge two execution fields. Assumes that they are fields for the same
+-- response key on the same type (i.e. that they are fields we would actually
+-- rationally want to merge).
+mergeFields :: Eq value => ExecutionField value -> ExecutionField value -> Validation (ExecutionField value)
+mergeFields field1 field2 = do
+  unless (name field1 == name field2) $ throwE (MismatchedNames (name field1) (name field2))
+  unless (arguments field1 == arguments field2) $ throwE (MismatchedArguments (name field1))
+  case (subSelectionSet field1, subSelectionSet field2) of
+    (Nothing, Nothing) ->
+      pure ExecutionField { name = name field1
+                          , arguments = arguments field1
+                          , subSelectionSet = Nothing
+                          }
+    (Just ss1, Just ss2) -> do
+      mergedSet <- mergeSelectionSets ss1 ss2
+      pure ExecutionField { name = name field1
+                          , arguments = arguments field1
+                          , subSelectionSet = Just mergedSet
+                          }
+    _ -> throwE (IncompatibleFields (name field1))
```

Comment:
Ah I see, how annoying! I wonder what a query language looks like that doesn't allow this .. probably doesn't have fragments!